python/rikai/spark/sql/codegen/base.py

```diff
@@ -63,8 +63,8 @@ def resolve(self, spec: "ModelSpec"):
         "transforms": {
             "type": "object",
             "properties": {
-                "pre": {"type": "string"},
-                "post": {"type": "string"},
+                "pre": {"type": ["string", "null"]},
```

Comment:
the schema only declares schema, version, and model to be required. So the whole `transforms` part is optional

Reply:
https://github.com/eto-ai/rikai/blob/c20aead79bd2525ce9d75140cd42871765abbce8/python/rikai/spark/sql/codegen/mlflow_registry.py#L227-L232

https://github.com/eto-ai/rikai/blob/c20aead79bd2525ce9d75140cd42871765abbce8/python/rikai/spark/sql/codegen/mlflow_registry.py#L239-L251

Without "null", json schema check will fail. We need the None value to by-pass pre/post processing for performance concerns.

Reply:
Could you clarify what performance concern it is?  I am confused that

```
if transforms.pre_processing is not None:
   pre_processing(data)

# vs

pre_processing = lambda x: x
pre_processing(data)
```
will have meaningful performance difference down the line.

Reply:
For example:

```
def demo(pre_processing: Optional[Callable], data):
    ...
    if pre_processing is not None:
        pre_processing(data)
    ...
```

When pre_processing is None, pre_processing on the possible large data set: `data` will not be performed.

But when pre_processing is identity, applying identity processing on the possible large data set `data` should be time-consuming.

Reply:
> Without "null", json schema check will fail. We need the None value to by-pass pre/post processing for performance concerns.

A alternative solution is to remove the key (pre/post processors) if the value is None in `get_model_conf`.

Reply:
> But when pre_processing is the identity function, applying identity processing on the possible large data set data should be time-consuming.

i think this was my question.  Given that this `pre_processing` function is `lambda x: x`, which just passed a reference of a dataset (i.e., tensor or large numpy array) in and out of the function, I am not certain that it will have measurable performance difference.